libpius/path.py

```diff
@@ -0,0 +1,150 @@
+                and not os.path.isdir(fn))
+
+    # If we're given a path with a directory part, look it up directly rather
+    # than referring to PATH directories. This includes checking relative to the
+    # current directory, e.g. ./script
+    if os.path.dirname(cmd):
+        if _access_check(cmd, mode):
+            return cmd
+        return None
+
+    if path is None:
+        path = os.environ.get("PATH", os.defpath)
+    if not path:
+        return None
+    path = path.split(os.pathsep)
+
+    if sys.platform == "win32":
+        # The current directory takes precedence on Windows.
+        if not os.curdir in path:
+            path.insert(0, os.curdir)
+
+        # PATHEXT is necessary to check on Windows.
+        pathext = os.environ.get("PATHEXT", "").split(os.pathsep)
+        # See if the given file matches any of the expected path extensions.
+        # This will allow us to short circuit when given "python.exe".
+        # If it does match, only test that one, otherwise we have to try
+        # others.
+        if any(cmd.lower().endswith(ext.lower()) for ext in pathext):
+            files = [cmd]
+        else:
+            files = [cmd + ext for ext in pathext]
+    else:
+        # On other platforms you don't have things like PATHEXT to tell you
+        # what file suffixes are executable, so just pass on cmd as-is.
+        files = [cmd]
+
+    seen = set()
+    for dir in path:
+        normdir = os.path.normcase(dir)
+        if not normdir in seen:
+            seen.add(normdir)
+            for thefile in files:
+                name = os.path.join(dir, thefile)
+                if _access_check(name, mode):
+                    return name
+    return None
+#end of "Which" function
+
+def gpg_path():
+    if sys.platform = "win32":
+        gpg = which('gpg2')
+    elif sys.platform = "darwin":
+        #joining all possible paths for location of mac binaries
+        BIN_PATHS = os.path.join(MAC_BIN_PATHS, LINUX_BIN_PATHS)
+        gpg = which('gpg2', path=BIN_PATHS)
+    elif sys.platform = "linux":
+        gpg = which('gpg2', path=LINUX_BIN_PATHS)
+    if gpg == "":
+        print("GPG2 could not be found! Do you have the correct permissions?")
+        sys.exit(1)
+    else:
+        return gpg
+
+def get_home():
+    if sys.platform == "win32":
+        return os.environ.get('APPDATA')
+    else:
+        return os.environ.get('HOME')
+
+def get_gpghome(usrhome):
+    if sys.platform == "win32":
+        return os.environ.get('GNUPGHOME', os.path.join(usrhome, 'gnupg'))
+    else:
+        return os.environ.get('GNUPGHOME', os.path.join(usrhome, '.gnupg'))
+
+def get_piushome(usrhome):
+    if sys.platform == "win32":
+        return os.path.join(usrhome, 'pius')
+    else:
+        #check to see if xdg home is a user env variable
+        #set it if it is to piushome
+        if os.environ.get('XDG_CONFIG_HOME') != "":
+            piushome = os.path.join(os.environ.get('XDG_CONFIG_HOME'), '.pius')
```

Comment:
if this doesn't exist, we should fall through to other cases, if they exist. Most users will have this defined, which means the moment they upgrade their old config will be lost.

Reply:
Correct there already is a fallback, the check to see if .config exists, also any files in the folder shouldn't be overwritten with an update/upgrade unless the owner permits that changes (most times it is just a config file and not actually deleting anything, it will defeat the purpose of and upgrade and would be more of a reinstallation)

Reply:
You're not falling through at the moment. If the variable is defined, but the file doesn't exist, you won't try any other locations, because you're doing `else`

Reply:
Yes I am, that if is checking XDG_CONFIG_HOME to see if it isn't empty, if it is then move to the elif (else if) which then checks for .config and if you are able to access it, so on and so forth. Trust me I wouldn't just check one thing and move directly to a fall back, i will try all possibilities before going to fallback

Reply:
Right, but many window managers DO define it, even though people ALREADY have a config in ~/.piusrc which means this code will NEVER see that config file.

Which is broken.

Reply:
I will fix it to use pius directory check if first time run.